README.md
# ELI, Expression Libre Incongrue

## About

ELI is a javascript widget that permits to show a GNU Social timeline. It can display either a user timeline or a group timeline.

We usually call it a *ELI's Badge*, a *ELI's Timeline* or just **ELI**.

I suggest you to checkout [our official demo page](http://eli.o9.re/ "Discover ELI in action") to have an idea about what ELI can do.

[![Demo](https://img.shields.io/badge/demo-official-green.svg)](http://eli.o9.re/ "Discover ELI in action")
![Javascript](https://img.shields.io/badge/code-javascript-bb0000.svg)

## Overview

ELI retrieves some data from GNU Social network and create a *Badge*.

You're so able to:

  * display either user or group timeline
  * display its avatar/logo
  * choose the domain from which you fetch statuses
  * set a limit of displayed statuses
  * display statuses ' attachments
  * display time and location from each status
  * choose time/location displayed label
  * display your main page even if ELI don't finish to retrieve data

![User Timeline](eli_attachment.png)   ![Group Timeline](eli_group.png)

Check local demo using [**minimal.html**](minimal.html) page or try an interactive one with [**index.html**](index.html) page.

## Installation

Just:

  * copy the eli.js in your website directory
  * add this code in your page:

```html
    <div id="elitimeline">
      <p>Timeline activity...</p>
    </div>
```

Then add this code **at the end of your page before the `</body>` tag**:

```html
    <script type="text/javascript" src="./eli.js">
      var type = 'user'; // could be 'group' to follow a group
      var user = 'bl4n';
      var max = 5;
      var tag = 'elitimeline';
      var domain = 'myGNUSocialDomain.tld';
    </script>
```

Adapt domain content by your GNU Social Instance. For an example: **https://status.fsf.org**.

In order ELI widget to be more beautiful, add this in your CSS file or in **`</style>` tag**:

```css
#eli_widget {
width: 250px;
height: auto;
margin: 0;
padding: 0;
border: thin solid #eee;
border-radius: 12px;
box-shadow: 2px 1px 5px #000;
}

#eli_widget header {
text-align: left;
border-bottom: 5px solid #fb6104;
background-color: #43568e;
border-top-left-radius: 12px;
border-top-right-radius: 12px;
color: #ffffff;
}

#eli_widget header img {
margin: 5px;
background-color: #ffffff;
border-top-left-radius: 12px;
float: left;
}

#eli_widget header p {
margin: 0;
line-height: 60px;
}

#eli_widget article {
display: block;
margin-bottom: 5px;
padding: 5px;
border-bottom: thin solid #eeeeee;
overflow: hidden; /* For too long link: hide but enable click on it */
}

#eli_widget article img {
margin: 5px;
max-width: 230px;
}

#eli_widget article p {
margin-top: 0px;
}

.tooltip {position: relative;}
.tooltip span {display: none;}
.tooltip:hover span {
display: block;
position: absolute;
left: 0; top: 10%;
margin: 20px 0 0;
width: 200px;
color: #4D4D4C;
border: thin solid #eeeeee;
padding: 4px;
background: white;
}
```

A more fully CSS example is available in **style.css** file.

## Configuration

When adding ELI's javascript declaration, you can use these variables:

  * type: Either **user** OR **group**. The last one is used when you want to follow a group
  * user: the name of the user (or the group) you want to follow
  * max: maximum displayed statuses. By default GNU Social seems to give only 20 statuses
  * tag: tagname ELI will use to replace its content and display the badge
  * domain: your GNU Social instance URL
  * timeLabel: Label that appears on time entry
  * locationLabel: Label that appears on location entry

## Contribution

[ELI is available on Github](http://github.com/blankoworld/eli).

**eli.js** is a minimal version of ELI. To have a more readable version, and to develop it, we recommand you to use **eli.max.js** file.

After having improved and tested ELI's code, **don't forget to minify your code** by using http://javascript-minifier.com/:

  * Copy/paste the content of eli.max.js
  * Delete the license lines, and these functions:
    * loadForm()
    * loadConfig()
  * Click on **Minify**
  * Copy the result and replace eli.js file content

Then publish your result on Github.

## License

This program is under the [WTF Public License](http://sam.zoy.org/wtfpl/COPYING 'Read more about the WTF Public License').

## Contributors

  * [LubuWest](https://github.com/LubuWest)
  * [mart-e](https://github.com/mart-e)
  * [Olivier DOSSMANN](https://github.com/blankoworld)


CHANGELOG
Current version (0.4-trunk):
  - Change official webpage link
  - Example quitter.se don't work while cross-site origin problem. Use fsf instead.
  - improve README.md file by adding Overview section, Contributors, etc.
  - new CHANGELOG file
  - asynchronous calls from the page to XML files
  - common function to set configuration from initial script and form
  - Change coding style: indentation, whitespaces, more functions, beautify code, use createElement() instead of HTML code

Version 0.3:
  - improve documentation with screenshots
  - identi.ca is closed. Use another social network: quitter.se
  - new "domain" variable to select the domain
  - display statuses ' attachments in timeline with a link to them
  - delete "div" tag from statuses to be compatible with timeline display
  - add a link on pictures from user/group to check their timeline
  - add time and location data from each status
  - add 2 variables to adapt displayed text: posttime_label and location_label
  - improve display of homepage

Version 0.2:
  - extract ELI's javascript code to an external file
  - add an external stylesheet
  - fetch picture from user/group to display it
  - improve documentation
  - make a minimal javascript version

Version 0.1.1:
  - add a form to change variables

Version 0.1:
  - can display user or group
  - set a "max" variable to choose how many statuses to display
  - README file


eli.js
var type="group",user="fsf",max=5,tag="elitimeline",domain="https://status.fsf.org",locationLabel="Location:",timeLabel="Post time:",api="/api/",user_api="users/show/",user_rss="statuses/user_timeline/",group_api="statusnet/groups/show/",group_rss="statusnet/groups/timeline/",api_url="",api_rss_url="",img_name="",template="",widget="",setConfig=function(e,t,a){api_url=e+api+user_api,api_rss_url=e+api+user_rss,img_name="profile_image_url","group"===a&&(api_url=e+api+group_api,api_rss_url=e+api+group_rss,img_name="stream_logo"),url=api_url+t+".xml"},getUrl=function(e,t){window.XMLHttpRequest?req=new XMLHttpRequest:req=new ActiveXObject("Microsoft.XMLHTTP"),req.open("GET",e),req.addEventListener("load",function(){req.status>=200&&req.status<400&&t(req.responseXML)}),req.send(null)},displayResult=function(){getUrl(url,createTimeline)},getTargetStatuses=function(e){for(var t=e.querySelector("statuses").getElementsByTagName("status"),a=0,r=0;r<t.length;r++){var n=t[r],i="",l="";try{l=n.querySelector("qvitter_delete_notice").textContent}catch(o){}if("true"!==l){var s=createMemberInfoBloc(n);try{i=n.getElementsByTagNameNS("*","html").item(0).firstChild.data,i=i.replace(i.slice(i.indexOf("<div"),i.indexOf("div>")+4),"")}catch(o){}var c="";try{var m=n.querySelector("attachments enclosure");if(["image/jpeg","image/gif","image/png","image/svg"].indexOf(m.getAttribute("mimetype"))>=0){var p=m.getAttribute("url");c=document.createElement("a"),c.target="_blank",c.href=p;var d=document.createElement("img");d.src=p,d.alt="Attachment",c.appendChild(d)}}catch(o){}var u="";try{var g=new Date(n.querySelector("created_at").textContent).toLocaleString();u=document.createElement("span"),u.className="tooltip";var h=document.createElement("span");h.appendChild(document.createTextNode(timeLabel+" "+g)),u.appendChild(document.createTextNode(" ⌛ ")),u.appendChild(h)}catch(o){}var _="";try{var f=n.querySelector("user location").textContent;_=document.createElement("span"),_.className="tooltip",_.appendChild(document.createTextNode(" ⌂ "));var v=document.createElement("span");v.textContent=locationLabel+" "+f,_.appendChild(v)}catch(o){}var C=document.createElement("article"),y="";""!==s&&(y+='<span class="groupmember">'+s.innerHTML+"</span>"),""!==u&&(y+='<span class="tooltip">'+u.innerHTML+"</span>"),""!==_&&(y+='<span class="tooltip">'+_.innerHTML+"</span>"),y+="<br>",y+=i,""!==c&&(y+=c.innerHTML),y+="",C.innerHTML=y,widget.insertBefore(C,templateFooter),a++}a===max&&(r=t.length)}},createTemplate=function(e){template=document.createElement("div"),template.id=tag,widget=document.createElement("div"),widget.id="eli_widget";var t=document.createElement("header");if(""!==e){var a=document.createElement("a");a.target="_blank",a.href=domain+"/"+user;var r=document.createElement("img");r.src=e,r.title="Avatar",r.alt=user,a.appendChild(r),t.appendChild(a)}var n=document.createElement("p");n.textContent=" "+user,t.appendChild(n),widget.appendChild(t),templateFooter=document.createElement("footer"),widget.appendChild(templateFooter)},createTimeline=function(e){var t=(e.getElementsByTagName(type),"");try{t=e.querySelector(img_name).textContent}catch(a){}try{targetId=e.querySelector("id").textContent}catch(a){throw new Error(alert("Can't fetch social network info."))}createTemplate(t),rss_url=api_rss_url+targetId+".xml",getUrl(rss_url,getTargetStatuses),template.appendChild(widget);var r=document.getElementById(tag);parent=r.parentNode,parent.replaceChild(template,r)},createMemberInfoBloc=function(e){var t="";if("group"===type)try{var a="@"+e.querySelector("screen_name").textContent,r=e.querySelector("name").textContent,n=e.querySelector("profile_image_url_https").textContent,i=e.getElementsByTagNameNS("*","profile_url").item(0).firstChild.data;t=document.createElement("span"),t.className="groupmember";var l=document.createElement("a");l.className="tooltip",l.target="_blank",l.href=i;var o=document.createElement("img");o.alt="Profile",o.align="left",o.src=n,l.appendChild(o),l.appendChild(document.createTextNode(a));var s=document.createElement("span");s.textContent=r,l.appendChild(s),t.appendChild(l)}catch(c){}return t};setConfig(domain,user,type),window.onload=displayResult;


eli.max.js
/* This program is free software. It comes without any warranty, to
 * the extent permitted by applicable law. You can redistribute it
 * and/or modify it under the terms of the Do What The Fuck You Want
 * To Public License, Version 2, as published by Sam Hocevar. See
 * http://sam.zoy.org/wtfpl/COPYING for more details. */

var loadForm = document.getElementById('loadForm');
loadForm.addEventListener('submit', function(e) {
    e.preventDefault();
    loadConfig();
}, true);

// Change variables here.
var type = 'group'; // could be 'group' to follow a group
var user = 'fsf';
var max = 5;
var tag = 'elitimeline';
var domain = 'https://status.fsf.org';
var locationLabel = 'Location:';
var timeLabel = 'Post time:';

// Network access API url
var api = '/api/';
var user_api = 'users/show/';
var user_rss = 'statuses/user_timeline/';
var group_api = 'statusnet/groups/show/';
var group_rss = 'statusnet/groups/timeline/';

// Other variables
var api_url = "",
    api_rss_url = "",
    img_name = "",
    template = "",
    widget = "";

var setConfig = function setConfig(domain, user, type) {
    api_url = domain + api + user_api;
    api_rss_url = domain + api + user_rss;
    img_name = "profile_image_url";
    if (type === "group") {
        api_url = domain + api + group_api;
        api_rss_url = domain + api + group_rss;
        img_name = "stream_logo";
    }
    url = api_url + user + ".xml";
};

var loadConfig = function loadConfig() {
    // fetch user choices
    domain = document.getElementById('profile_domain').value;
    user = document.getElementById('profile_name').value;
    var choosenType = document.getElementById('profile_type');
    type = choosenType.options[profile_type.selectedIndex].value;
    // change current configuration
    setConfig(domain, user, type);
    // display the result
    displayResult();
};

var getUrl = function getUrl(url, method) {
    if (window.XMLHttpRequest) {
        // code for IE7+, Firefox, Chrome, Opera, Safari
        req = new XMLHttpRequest();
    } else {
        // code for IE6, IE5
        req = new ActiveXObject("Microsoft.XMLHTTP");
    }
    req.open("GET", url);
    req.addEventListener("load", function () {
        if (req.status >= 200 && req.status < 400) {
            method(req.responseXML);
        }
    });
    req.send(null);
};

var displayResult = function displayResult() {
    // get main info
    getUrl(url, createTimeline);
};

var getTargetStatuses = function getTargetStatuses(source) {
    // get statuses from source
    var statuses = source.querySelector('statuses').getElementsByTagName('status');

    // browse statuses but no more than "max"
    var displayedItemCounter = 0;
    for (var cursor = 0; cursor < statuses.length; cursor++) {
        var status = statuses[cursor];
        var content = "";

        // Check that this status can be displayed. If status deleted, do not do anything
        var statusDeleted = "";
        try {
            statusDeleted = status.querySelector('qvitter_delete_notice').textContent;
        } catch (e) {}
        // Check status and add it to timeline only if it was not deleted
        if (statusDeleted !== 'true') {
            // Add a bloc about member that contains images, names and link
            var memberInfo = createMemberInfoBloc(status);

            // get status real content (in html)
            try {
                content = status.getElementsByTagNameNS('*', 'html').item(0).firstChild.data;
                content = content.replace(content.slice(content.indexOf('<div'), content.indexOf('div>') + 4), '');
            } catch (e) {}

            // If some picture with a link, get it.
            var imageElt = "";
            try {
                var enclosure = status.querySelector('attachments enclosure');
                if (['image/jpeg', 'image/gif', 'image/png', 'image/svg'].indexOf(enclosure.getAttribute("mimetype")) >= 0) {
                    var imageUrl = enclosure.getAttribute("url");
                    imageElt = document.createElement("a");
                    imageElt.target = "_blank";
                    imageElt.href = imageUrl;
                    var image = document.createElement("img");
                    image.src = imageUrl;
                    image.alt = "Attachment";
                    imageElt.appendChild(image);
                }
            } catch (e) {}

            // Status created date
            var timeElt = "";
            try {
                var time = new Date(status.querySelector('created_at').textContent).toLocaleString();
                timeElt = document.createElement("span");
                timeElt.className = "tooltip";
                var labelElt = document.createElement("span");
                labelElt.appendChild(document.createTextNode(timeLabel + " " + time));
                timeElt.appendChild(document.createTextNode(" ⌛ "));
                timeElt.appendChild(labelElt);
            } catch (e) {}

            // Status location
            var locationElt = "";
            try {
                var location = status.querySelector('user location').textContent;
                locationElt = document.createElement("span");
                locationElt.className = "tooltip";
                locationElt.appendChild(document.createTextNode(" ⌂ "));
                var labelLocationElt = document.createElement("span");
                labelLocationElt.textContent = locationLabel + " " + location;
                locationElt.appendChild(labelLocationElt);
            } catch (e) {}

            // Add status on timeline
            var item = document.createElement("article");
            var htmlContent = "";
            if (memberInfo !== "") {
                htmlContent += "<span class=\"groupmember\">" + memberInfo.innerHTML + "</span>";
            }
            if (timeElt !== "") {
                htmlContent += "<span class=\"tooltip\">" + timeElt.innerHTML + "</span>";
            }
            if (locationElt !== "") {
                htmlContent += "<span class=\"tooltip\">" + locationElt.innerHTML + "</span>";
            }
            htmlContent += "<br>";
            htmlContent += content;
            if (imageElt !== "") {
                htmlContent += imageElt.innerHTML;
            }
            htmlContent += "";
            item.innerHTML = htmlContent;
            widget.insertBefore(item, templateFooter);

            // increment number of item displayed on screen
            displayedItemCounter++;
        }

        // Hack to break the loop: set cursor to its maximum.
        if (displayedItemCounter === max) {
            cursor = statuses.length;
        }
    }
};

var createTemplate = function createTemplate(image) {
    // Initialization
    template = document.createElement("div");
    template.id = tag;
    widget = document.createElement("div");
    widget.id = "eli_widget";
    // Add header with the username and a link to its social network
    var templateHeader = document.createElement("header");
    if (image !== "") {
        var networkLink = document.createElement("a");
        networkLink.target = "_blank";
        networkLink.href = domain + "/" + user;
        var targetImage = document.createElement("img");
        targetImage.src = image;
        targetImage.title = "Avatar";
        targetImage.alt = user;
        networkLink.appendChild(targetImage);
        templateHeader.appendChild(networkLink);
    }
    var content = document.createElement("p");
    content.textContent = "\u00A0" + user;  // add a space before target
    templateHeader.appendChild(content);
    widget.appendChild(templateHeader);
    // Finally add footer
    templateFooter = document.createElement("footer");
    widget.appendChild(templateFooter);
};

var createTimeline = function createTimeline(source) {
    // get info about target
    var targetInfo = source.getElementsByTagName(type);
    // get image source
    var imageSource = "";
    try {
        imageSource = source.querySelector(img_name).textContent;
    } catch (e) {}

    // get target ID to fetch RSS feed (for an example with ID 3, RSS feed is 3.xml)
    try {
        targetId = source.querySelector("id").textContent;
    } catch (e) {
        throw new Error(alert("Can't fetch social network info."));
    }

    // create template
    createTemplate(imageSource);

    // fetch RSS feed
    rss_url = api_rss_url + targetId + ".xml";
    getUrl(rss_url, getTargetStatuses);

    // replace timeline by new content
    template.appendChild(widget);
    var timeline = document.getElementById(tag);
    parent = timeline.parentNode;
    parent.replaceChild(template, timeline);
};

var createMemberInfoBloc = function createMemberInfoBloc(text) {
    // only create this bloc if type is group
    var infoBloc = "";
    if (type === "group") {
        try {
            // get data
            var pseudo = "@" + text.querySelector('screen_name').textContent;
            var name = text.querySelector('name').textContent;
            var image = text.querySelector('profile_image_url_https').textContent;
            var link = text.getElementsByTagNameNS('*', 'profile_url').item(0).firstChild.data;
            infoBloc = document.createElement("span");
            // then start creating bloc
            infoBloc.className = "groupmember";
            var linkElt = document.createElement("a");
            linkElt.className = "tooltip";
            linkElt.target = "_blank";
            linkElt.href = link;
            // add image, pseudo, then name into the link
            var imageElt = document.createElement("img");
            imageElt.alt = "Profile";
            imageElt.align = "left"; // TODO: Add this directive into CSS (add imageElt.classList.add("member"))
            imageElt.src = image;
            linkElt.appendChild(imageElt);
            linkElt.appendChild(document.createTextNode(pseudo));
            var nameElt = document.createElement("span");
            nameElt.textContent = name;
            linkElt.appendChild(nameElt);
            // finally add link to the bloc
            infoBloc.appendChild(linkElt);
        } catch (e) {}
    }
    return infoBloc;
};

// Set default configuration
setConfig(domain, user, type);

window.onload = displayResult;
